Fix auto-related FE component [WEB-2699]

web-dev-trev opened the pull request.

zachgarwood commented on resources/views/site/shared/_featuredRelated.blade.php:
```diff
@@ -17,9 +17,30 @@
+    $isFeatured = false;
+    $featuredRelated = collect($item->getFeaturedRelated())->filter()->map(function ($featuredRelated) {
+        return $featuredRelated;
+    });
```
I don't quite understand what's happening here. The `filter()` will remove the `null` items, but the `map()` seems to just return every item without modification.

zachgarwood replied:
I see you removed the `filter()`, but what's the purpose of the `map()` that just returns the item?

nikhiltri replied:
I think you can use the Laravel Collections `[pluck](https://laravel.com/docs/10.x/collections#method-pluck)` method to do this more efficiently. 

```
collect($item->getFeaturedRelated())->pluck('item');
```

zachgarwood commented on resources/views/site/shared/_featuredRelated.blade.php:
```diff
@@ -17,9 +17,30 @@
+
+    $featuredRelatedIds = $featuredRelated->pluck('id');
+
+    dump($featuredRelated);
```
Remove debugging code.

zachgarwood reviewed (changes requested):
A couple things to address

web-dev-trev commented:
@zachgarwood @nikhiltri I added some more changes to auto-related and addressed Zach's initial comments. This is my first go at using usort() so hopefully everything makes sense. I left comments throughout to explain what's happening but let me know if anything else looks weird.

nikhiltri commented on app/Models/Behaviors/HasAutoRelated.php:
```diff
@@ -22,17 +27,77 @@ public function related($id)
+        ];
+
+        // Set scope to only include sidebar items that can be related to
+        $relatedSidebarItems = DB::table('related')
+            ->where('browser_name', 'sidebar_items')
+            ->where('related_id', $id)
+            ->where('related_type', $modelName)
+            ->whereIn('subject_type', $editorialTypes)
+            ->get();
```
Rather than constructing a DB query here, use the A17\Twill\Models\RelatedItem model and the Eloquent methods for retrieving models: https://laravel.com/docs/10.x/eloquent#retrieving-models

nikhiltri commented on app/Models/Behaviors/HasFeaturedRelated.php:
```diff
@@ -43,14 +43,6 @@ public function getFeaturedRelated()
         if (!$this->selectedFeaturedRelateds) {
             $relatedItems = $this->getCustomRelatedItems();
 
-            if (config('aic.show_default_related_items')) {
-                if (($this->showDefaultRelatedItems ?? false) && $relatedItems->count() < 1) {
-                    $relatedItems = $this->getDefaultRelatedItems($relatedItems);
-                    $relatedItems = $relatedItems->slice(0, $this->getTargetItemCount());
-                    $this->sidebarContainsDefaultRelated = true;
-                }
-            }
-
```
👋🏾

nikhiltri reviewed (changes requested):
Some minor changes, but looks good overall!

zachgarwood commented on app/Models/Behaviors/HasAutoRelated.php:
```diff
@@ -22,17 +27,77 @@ public function related($id)
+            }
+
+            return 1;
+        });
```
👍

nikhiltri commented on resources/views/admin/partials/featured-related.blade.php:
```diff
@@ -36,4 +36,30 @@
         ],
     ])
+
+    @php
+        $autoRelated = collect($item->related($item->id))->unique('id')->filter();
+
+        $featuredRelated = collect($item->getFeaturedRelated())->pluck('item');
+        $featuredRelatedIds = $featuredRelated->pluck('id');
+
+        // Remove featured related items from auto related items
+        if ($featuredRelatedIds->isNotEmpty()) {
+            $autoRelated = $autoRelated->reject(function ($relatedItem) use ($featuredRelatedIds) {
+                return ($relatedItem !== null && ($featuredRelatedIds->contains($relatedItem->id) || $featuredRelatedIds->contains($relatedItem->datahub_id)));
+            });
+        }
+    @endphp
```
Move this into the `formData` methods of the Admin Controllers in which this is applicable and pass in `$autoRelated`. Then on the next line below check `isset($autoRelated) && count($autoRelated) > 0`.

In the MVC model of coding, logic like this shouldn't live in views.

nikhiltri commented on resources/views/site/shared/_featuredRelated.blade.php:
```diff
@@ -17,9 +17,26 @@
     if (($behavior ?? '') === 'relatedSidebar') {
         $listingVariation .= ' m-listing--dynamic';
     }
+
+    // Get featured related items
+    $isFeatured = false;
+    $featuredRelated = collect($item->getFeaturedRelated())->pluck('item');
+
+    $featuredRelatedIds = $featuredRelated->pluck('id');
+
+    // Get auto related items & evaluate if they are featured
+
+    $autoRelated = collect($item->related($item->id))->unique('id')->filter();
+
+    // Remove featured related items from auto related items
+    if ($featuredRelatedIds->isNotEmpty()) {
+        $autoRelated = $autoRelated->reject(function ($relatedItem) use ($featuredRelatedIds) {
+            return ($relatedItem !== null && ($featuredRelatedIds->contains($relatedItem->id) || $featuredRelatedIds->contains($relatedItem->datahub_id)));
+        });
+    }
```
Same as previous comment, move this into the `show()` methods of the applicable controllers and pass an `$autoRelated` array to the view.

nikhiltri reviewed (changes requested):
A few minor notes about code organization. Really close!

nikhiltri commented on app/Http/Controllers/Admin/EventController.php:
```diff
@@ -109,7 +109,20 @@ protected function formData($request)
+        }
+
         return [
+            'autoRelated' => $autoRelated,
```
Should `$featureRelated` also be passed here?

nikhiltri commented on app/Http/Controllers/Admin/ExhibitionController.php:
```diff
@@ -93,7 +93,20 @@ protected function formData($request)
+        }
+
         return [
+            'autoRelated' => $autoRelated,
```
Should `$featureRelated` also be passed here?

nikhiltri commented on app/Http/Controllers/Admin/HighlightController.php:
```diff
@@ -61,9 +61,23 @@ protected function indexData($request)
+        }
+
         return [
+            'autoRelated' => $autoRelated,
```
Should `$featureRelated` also be passed here?

nikhiltri commented on app/Http/Controllers/Admin/EventController.php:
```diff
@@ -110,6 +109,7 @@ protected function formData($request)
         $baseUrl = '//' . config('app.url') . '/events/' . $item->id . '/';
 
         return [
+            'autoRelated' => $this->getAutoRelated($item),
```
Should `$featureRelated` also be passed here?

nikhiltri commented on app/Http/Controllers/Admin/ExhibitionController.php:
```diff
@@ -94,6 +94,7 @@ protected function formData($request)
         $baseUrl = '//' . config('app.url') . '/exhibitions/' . $item->datahub_id . '/';
 
         return [
+            'autoRelated' => $this->getAutoRelated($item),
```
Should `$featureRelated` also be passed here?

nikhiltri commented on app/Http/Controllers/Admin/HighlightController.php:
```diff
@@ -61,9 +60,11 @@ protected function indexData($request)
         $baseUrl = '//' . config('app.url') . '/highlights/' . request('highlight') . '/';
 
         return [
+            'autoRelated' => $this->getAutoRelated($item),
```
Should `$featureRelated` also be passed here?

nikhiltri reviewed (approved):
Nice work! ⭐
